### ISSUE TITLE
LoadBalancer: support instance selection based on specific actuator metrics

### ISSUE THREAD

**Issue by kmandalas** [spring-cloud/spring-cloud-commons#756]
Provide more fine-grained load balancing rules based on actuator metrics like `system.cpu.usage`, `jvm.memory.usage` etc. Could be something configurable based on a specific metric and a corresponding threshold value. If this value is exceeded the balancer will prefer instances below this value. Otherwise default behavior will be applied.

Alternatively I assume this can be achieved by implementing a custom `HealthIndicator` but a more out-of-the-box configuration-only capability could be very useful.

**Comment by spencergibb** [comment]
#601 is along those lines. It's fairly hard to do since you would want those metrics from other instances, how do you report that the load balancer.

**Comment by kmandalas** [comment]
@spencergibb so the only solution would be to implement a custom `HealthIndicator`?

**Comment by spencergibb** [comment]
that's one option, yes.

**Comment by kmandalas** [comment]
@spencergibb I think that custom `HealthIndicator` should be avoided though since it would show a status [DOWN] if some instance is under load exceeding some threshold and this is will not be accurate. I guess implementing a custom `LoadBalancerClientFilter` or `ReactiveLoadBalancerClientFilter` could be the only alternative at the moment.

**Comment by kmandalas** [comment]
@spencergibb I was working on a custom `ReactorServiceInstanceLoadBalancer ` at the moment cause I want to achieve a "least_conn" behavior (similarly to NGNIX or HA_PROXY). This is usefult in many cases one of which is when you need to load balance WebSocket connections from the Spring Cloud Gateway to multiple instances of WebSocket servers. Do you think this would of any interest to prepare a PR?

**Comment by OlgaMaciaszek** [comment]
@kmandalas Definitely, if you come up with something, do submit a PR. You might also want to keep an eye on what is happening with the following issues: https://github.com/spring-cloud/spring-cloud-commons/issues/675 (currently a PR in review - introduces possibilities to propagate load-balanced call data and to run a callback method after a load-balanced call has been completed; probably best to base your changes on that) and https://github.com/spring-cloud/spring-cloud-commons/issues/674 (planning to work on adding in micrometer here).
